### PR TITLE
Add extended data to calculated properties

### PR DESCRIPTION
**Pull request by JonasDov**
itwinjs-core: https://github.com/iTwin/itwinjs-core/pull/7164

### CODE REVIEW

File: iModelCore/ECPresentation/Source/Content/Content.cpp [hunk 1]
```diff
@@ -1425,6 +1425,12 @@ void ContentSetItem::AddUsersExtendedData(Utf8CP key, ECValueCR value)
     m_extendedData.AddMember(rapidjson::Value(key, m_extendedData.GetAllocator()), ValueHelpers::GetJsonFromECValue(value, "", &m_extendedData.GetAllocator()), m_extendedData.GetAllocator());
     }
 
+
+void ContentDescriptor::CalculatedPropertyField::AddExtendedData(Utf8CP key, ECValueCR value)
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/Content/Content.cpp:1429]:
```suggestion
/*---------------------------------------------------------------------------------**//**
* @bsimethod
+---------------+---------------+---------------+---------------+---------------+------*/
void ContentDescriptor::CalculatedPropertyField::AddExtendedData(Utf8CP key, ECValueCR value)
```

File: iModelCore/ECPresentation/Source/Content/ContentDescriptorBuilder.cpp [hunk 2]
```diff
@@ -387,6 +387,23 @@ struct ContentPropertiesAppender : ContentSpecificationsHandler::PropertyAppende
...
         if (nullptr != spec.GetEditor())
             field->SetEditor(ContentFieldEditor::FromSpec(*spec.GetEditor()));
+
+        if (spec.GetExtendedDataMap().size() == 0)
+            return field;
+
+        ECExpressionContextsProvider::ContextParametersBase params(m_context.GetConnection(), m_context.GetRulesetVariables(), m_context.GetUsedVariablesListener());
+        ExpressionContextPtr expressionContext = ECExpressionContextsProvider::GetRulesEngineRootContext(params);
+
+        ECValue value;
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/Content/ContentDescriptorBuilder.cpp:397]: looks unnecessary

File: iModelCore/ECPresentation/Source/Rules/CalculatedPropertiesSpecification.cpp [hunk 3]
```diff
@@ -116,6 +119,25 @@ bool CalculatedPropertiesSpecification::_ReadJson(BeJsConst json)
     if (json.hasMember(CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_EDITOR))
         m_editor = CommonToolsInternal::LoadRuleFromJson<PropertyEditorSpecification>(json[CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_EDITOR]);
 
+    if (!json.hasMember(CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_EXTENDEDDATA))
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/Rules/CalculatedPropertiesSpecification.cpp:122]: I think this condition should be inverted to `if (has member) { /* do stuff */ }`

File: iModelCore/ECPresentation/Source/Rules/CalculatedPropertiesSpecification.cpp [hunk 4]
```diff
@@ -137,6 +159,9 @@ void CalculatedPropertiesSpecification::_WriteJson(BeJsValue json) const
         m_editor->WriteJson(json[CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_EDITOR]);
     if (nullptr != m_categoryId)
         m_categoryId->WriteJson(json[CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_CATEGORYID]);
+
+    for (auto entry : m_extendedData)
+        json[CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_EXTENDEDDATA][entry.first] = entry.second;
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/Rules/CalculatedPropertiesSpecification.cpp:164]:
```suggestion
    for (auto const& entry : m_extendedData)
        json[CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_EXTENDEDDATA][entry.first] = entry.second;
```

File: iModelCore/ECPresentation/Source/Rules/CalculatedPropertiesSpecification.cpp [hunk 5]
```diff
@@ -187,5 +212,29 @@ MD5 CalculatedPropertiesSpecification::_ComputeHash() const
         ADD_STR_VALUE_TO_HASH(md5, CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_EDITOR, m_editor->GetHash());
     if (nullptr != m_categoryId)
         ADD_STR_VALUE_TO_HASH(md5, CALCULATED_PROPERTIES_SPECIFICATION_JSON_ATTRIBUTE_CATEGORYID, m_categoryId->GetHash());
+
+    for (auto entry : m_extendedData)
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/Rules/CalculatedPropertiesSpecification.cpp:216]:
```suggestion
    for (auto const& entry : m_extendedData)
```

File: iModelCore/ECPresentation/Tests/NonPublished/Integration/Content/ContentIntegrationTests.cpp [hunk 6]
```diff
@@ -3235,6 +3235,61 @@ TEST_F(RulesDrivenECPresentationManagerContentTests, ContentInstancesOfSpecificC
...
+)*");
+TEST_F(RulesDrivenECPresentationManagerContentTests, ContentInstancesOfSpecificClasses_CalculatedPropertiesExtendedData)
+    {
+    ECClassCP classA = GetClass("A");
+
+    // insert some instance
+    IECInstancePtr instance = RulesEngineTestHelpers::InsertInstance(s_project->GetECDb(), *classA, [](IECInstanceR instance)
+        {
+            instance.SetValue("Property", ECValue("Test"));
+        });
```
Comment (grigasp) [inline iModelCore/ECPresentation/Tests/NonPublished/Integration/Content/ContentIntegrationTests.cpp:3254]:
```suggestion
    IECInstancePtr instance = RulesEngineTestHelpers::InsertInstance(s_project->GetECDb(), *classA, [](IECInstanceR instance)
        {
        instance.SetValue("Property", ECValue("Test"));
        });

```

File: iModelCore/ECPresentation/Tests/NonPublished/Integration/Content/ContentIntegrationTests.cpp [hunk 7]
```diff
@@ -3235,6 +3235,61 @@ TEST_F(RulesDrivenECPresentationManagerContentTests, ContentInstancesOfSpecificC
...
+
+    // create the rule set
+    PresentationRuleSetPtr rules = PresentationRuleSet::CreateInstance(BeTest::GetNameOfCurrentTest());
+    m_locater->AddRuleSet(*rules);
+
+    ContentRuleP rule = new ContentRule("", 1, false);
+    ContentSpecificationP specification = new ContentInstancesOfSpecificClassesSpecification(1, "", classA->GetFullName(), false, false);
+    specification->AddCalculatedProperty(*new CalculatedPropertiesSpecification("instance id", 1000, "this.ECInstanceId"));
+    auto specificationWithExtendedData = new CalculatedPropertiesSpecification("extendedData", 1000, "1+1");
```
Comment (grigasp) [inline iModelCore/ECPresentation/Tests/NonPublished/Integration/Content/ContentIntegrationTests.cpp:3263]: I'd avoid naming this with the same name as the JSON key that we use. Instead, let's name it `with extended data`.

**Review by grigasp — APPROVED** [review]

File: iModelCore/ECPresentation/Source/Rules/CalculatedPropertiesSpecification.cpp (hunk 3 above)
Comment (JonasDov) [reply iModelCore/ECPresentation/Source/Rules/CalculatedPropertiesSpecification.cpp]: Changed

File: iModelCore/ECPresentation/Source/Content/ContentDescriptorBuilder.cpp (hunk 2 above)
Comment (JonasDov) [reply iModelCore/ECPresentation/Source/Content/ContentDescriptorBuilder.cpp]: Removed

**Comment by aruniverse** [comment]
@mergifyio backport release/4.9.x

**Pull request merged**